crates/nargo_cli/src/cli/compile_cmd.rs

```diff
@@ -49,9 +50,11 @@ pub(crate) fn run<B: Backend>(
     // If contracts is set we're compiling every function in a 'contract' rather than just 'main'.
     if args.contracts {
         let mut driver = setup_driver(backend, &config.program_dir)?;
-        let compiled_contracts = driver
-            .compile_contracts(&args.compile_options)
-            .map_err(|_| CliError::CompilationError)?;
+        let compiled_contracts =
+            driver.compile_contracts(&args.compile_options).map_err(|errors| {
+                reporter::report_all(driver.file_manager(), &errors, false);
```

Comment:
In the check command, we need to do 
```rs
let error_count = reporter::report_all(file_manager, &errs, compile_options.deny_warnings);
if error_count != 0 {
        reporter::finish_report(error_count);
        return Err(CliError::CompilationError);
}
```
I'm guessing the same will need to be done here?

Reply:
Right. I'm also realizing that this PR does not fix the case where we abort compilation if there were warnings but no errors. I'll add a commit to fix this as well.

Reply:
Yeah, the driver code is extremely gnarly since we can return success + warnings and error + warnings.

Reply:
Ok, I've pushed the slightly larger rework to avoid failing when only warnings are issued.

Instead of `(Option<T>, Vec<FileDiagnostic>)` I opted to go with `Result<(T, Vec<FileDiagnostic>), Vec<FileDiagnostic>>` where we return `(T, warnings)` on success, and on error the file diagnostic list will contain errors as well. To make it a bit easier to read I've added some type alises: `Result<(T, Warnings), ErrorsAndWarnings>`.